File: windows/WeChatExporter.Windows/Services/PdfExporter.cs
using System.Text;

namespace WeChatExporter.Services;

public static class PdfExporter
{
    public static string Write(string sourceDir, string contactName, string destinationDir)
    {
        System.IO.Directory.CreateDirectory(destinationDir);
        var safe = MarkdownExporter.Sanitize(contactName);
        var stamp = MarkdownExporter.FileStamp();
        var outPath = System.IO.Path.Combine(destinationDir, $"{safe}_{stamp}.pdf");
        var sidecarPath = System.IO.Path.Combine(destinationDir, $"{safe}_{stamp}_pdf内容.txt");

        var lines = LoadLines(sourceDir, contactName);
        System.IO.File.WriteAllBytes(outPath, BuildSimplePdf(lines));
        // The minimal PDF uses built-in Helvetica; keep a UTF-8 sidecar so Chinese text is preserved losslessly.
        System.IO.File.WriteAllText(sidecarPath, string.Join(Environment.NewLine, lines), Encoding.UTF8);
        return outPath;
    }

    private static List<string> LoadLines(string sourceDir, string contactName)
    {
        var lines = new List<string> { contactName, "由 WeChatExporter 导出", "" };
        var jsonPath = System.IO.Path.Combine(sourceDir, "chat.json");
        var txtPath = System.IO.Path.Combine(sourceDir, "chat.txt");
        if (System.IO.File.Exists(jsonPath))
        {
            foreach (var row in MarkdownExporter.ReadRows(jsonPath))
            {
                var source = row.TryGetProperty("message", out var nested) && nested.ValueKind == System.Text.Json.JsonValueKind.Object
                    ? nested
                    : row;
                var time = MarkdownExporter.GetString(row, "time", "timestamp_str")
                           ?? MarkdownExporter.GetString(source, "time", "timestamp_str")
                           ?? "";
                var sender = MarkdownExporter.GetString(row, "sender_display_name", "sender", "from")
                             ?? MarkdownExporter.GetString(source, "sender_display_name", "sender")
                             ?? "";
                var content = MarkdownExporter.GetString(row, "snippet", "content", "text")
                              ?? MarkdownExporter.GetString(source, "snippet", "content", "text")
                              ?? "";
                lines.Add($"[{time}] {sender}: {content}");
            }
        }
        else if (System.IO.File.Exists(txtPath))
        {
            lines.AddRange(System.IO.File.ReadAllLines(txtPath));
        }
        else
        {
            throw new InvalidOperationException("未找到聊天记录文件，无法生成 PDF");
        }
        return lines;
    }

    private static byte[] BuildSimplePdf(IReadOnlyList<string> lines)
    {
        const int pageWidth = 612;
        const int pageHeight = 792;
        const int margin = 48;
        const int lineHeight = 14;
        var linesPerPage = Math.Max(1, (pageHeight - margin * 2) / lineHeight);

        var pages = new List<List<string>>();
        for (var i = 0; i < Math.Max(lines.Count, 1); i += linesPerPage)
            pages.Add(lines.Skip(i).Take(linesPerPage).ToList());
        if (pages.Count == 0) pages.Add([]);

        var objects = new List<byte[]>();
        void Add(string text) => objects.Add(Encoding.ASCII.GetBytes(text));
        void AddBytes(byte[] data) => objects.Add(data);

        Add("<< /Type /Catalog /Pages 2 0 R >>\n");
        Add("PLACEHOLDER\n");
        Add("<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\n");

        var contentIds = new List<int>();
        foreach (var pageLines in pages)
        {
            var stream = new StringBuilder();
            stream.Append($"BT /F1 10 Tf {margin} {pageHeight - margin} Td {lineHeight} TL\n");
            for (var i = 0; i < pageLines.Count; i++)
            {
                var text = EscapePdf(AsciiOnly(pageLines[i]));
                stream.Append(i == 0 ? $"({text}) Tj\n" : $"T* ({text}) Tj\n");
            }
            stream.Append("ET");
            var body = Encoding.ASCII.GetBytes(stream.ToString());
            var obj = new List<byte>();
            obj.AddRange(Encoding.ASCII.GetBytes($"<< /Length {body.Length} >>\nstream\n"));
            obj.AddRange(body);
            obj.AddRange(Encoding.ASCII.GetBytes("\nendstream\n"));
            AddBytes(obj.ToArray());
            contentIds.Add(objects.Count);
        }

        var pageIds = new List<int>();
        foreach (var cid in contentIds)
        {
            Add($"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 {pageWidth} {pageHeight}] /Contents {cid} 0 R /Resources << /Font << /F1 3 0 R >> >> >>\n");
            pageIds.Add(objects.Count);
        }

        var kids = string.Join(" ", pageIds.Select(id => $"{id} 0 R"));
        objects[1] = Encoding.ASCII.GetBytes($"<< /Type /Pages /Kids [{kids}] /Count {pageIds.Count} >>\n");

        using var ms = new System.IO.MemoryStream();
        void WriteAscii(string text) => ms.Write(Encoding.ASCII.GetBytes(text));
        WriteAscii("%PDF-1.4\n");
        var offsets = new List<long> { 0 };
        for (var i = 0; i < objects.Count; i++)
        {
            offsets.Add(ms.Position);
            WriteAscii($"{i + 1} 0 obj\n");
            ms.Write(objects[i]);
            if (objects[i].Length == 0 || objects[i][^1] != (byte)'\n')
                WriteAscii("\n");
            WriteAscii("endobj\n");
        }

        var xref = ms.Position;
        WriteAscii($"xref\n0 {objects.Count + 1}\n0000000000 65535 f \n");
        foreach (var offset in offsets.Skip(1))
            WriteAscii($"{offset:0000000000} 00000 n \n");
        WriteAscii($"trailer\n<< /Size {objects.Count + 1} /Root 1 0 R >>\nstartxref\n{xref}\n%%EOF\n");
        return ms.ToArray();
    }

    private static string AsciiOnly(string text)
    {
        var chars = text
            .Take(110)
            .Select(ch => ch is '\\' or '(' or ')' ? ' ' : ch <= 0x7F ? ch : '?')
            .ToArray();
        return new string(chars);
    }

    private static string EscapePdf(string text) =>
        text.Replace("\\", "\\\\").Replace("(", "\\(").Replace(")", "\\)");
}

File: Sources/WeChatExporter/Services/PdfExporter.swift
import Foundation
#if canImport(AppKit)
import AppKit
import CoreText
#endif

/// 导出为 PDF（macOS 用 CoreText 排版支持中文；其它环境回退 ASCII PDF + 旁路 TXT）。
enum PdfExporter {
    static func write(from sourceDir: URL, contactName: String, into destinationDir: URL) throws -> URL {
        let lines = loadLines(from: sourceDir, contactName: contactName)
        let safe = sanitize(contactName)
        let stamp = timestamp()
        let out = destinationDir.appendingPathComponent("\(safe)_\(stamp).pdf")
        try FileManager.default.createDirectory(at: destinationDir, withIntermediateDirectories: true)

        #if canImport(AppKit)
        do {
            let data = try coreTextPDF(lines: lines)
            try data.write(to: out, options: .atomic)
            return out
        } catch {
            // fall through
        }
        #endif

        let data = fallbackASCIIPdf(lines: lines)
        try data.write(to: out, options: .atomic)
        let txtSide = destinationDir.appendingPathComponent("\(safe)_\(stamp)_pdf内容.txt")
        try lines.joined(separator: "\n").write(to: txtSide, atomically: true, encoding: .utf8)
        return out
    }

    #if canImport(AppKit)
    private static func coreTextPDF(lines: [String]) throws -> Data {
        let text = lines.joined(separator: "\n")
        let font = NSFont.systemFont(ofSize: 11)
        let paragraph = NSMutableParagraphStyle()
        paragraph.lineBreakMode = .byWordWrapping
        let attrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: NSColor.black,
            .paragraphStyle: paragraph,
        ]
        let attributed = NSAttributedString(string: text, attributes: attrs)

        let pageWidth: CGFloat = 612
        let pageHeight: CGFloat = 792
        let margin: CGFloat = 48
        let data = NSMutableData()
        var mediaBox = CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight)
        guard let consumer = CGDataConsumer(data: data as CFMutableData),
              let context = CGContext(consumer: consumer, mediaBox: &mediaBox, nil) else {
            throw AppError.exportFailed("无法创建 PDF 上下文")
        }

        let framesetter = CTFramesetterCreateWithAttributedString(attributed)
        var textPos = 0
        let total = attributed.length
        while textPos < total {
            context.beginPage(mediaBox: &mediaBox)
            context.textMatrix = .identity
            let frameRect = CGRect(
                x: margin,
                y: margin,
                width: pageWidth - margin * 2,
                height: pageHeight - margin * 2
            )
            let path = CGPath(rect: frameRect, transform: nil)
            let range = CFRange(location: textPos, length: total - textPos)
            let frame = CTFramesetterCreateFrame(framesetter, range, path, nil)
            CTFrameDraw(frame, context)
            let visible = CTFrameGetVisibleStringRange(frame)
            textPos += max(visible.length, 1)
            context.endPage()
        }
        context.closePDF()
        return data as Data
    }
    #endif

    private static func loadLines(from sourceDir: URL, contactName: String) -> [String] {
        var lines: [String] = ["\(contactName)", "由 WeChatExporter 导出", ""]
        let jsonURL = sourceDir.appendingPathComponent("chat.json")
        let txtURL = sourceDir.appendingPathComponent("chat.txt")
        if FileManager.default.fileExists(atPath: jsonURL.path),
           let data = try? Data(contentsOf: jsonURL),
           let root = try? JSONSerialization.jsonObject(with: data) {
            for row in messageRows(from: root) {
                let source = (row["message"] as? [String: Any]) ?? row
                let time = stringField(row, keys: ["time", "timestamp_str"])
                    ?? stringField(source, keys: ["time", "timestamp_str"])
                    ?? ""
                let sender = stringField(row, keys: ["sender_display_name", "sender", "from"])
                    ?? stringField(source, keys: ["sender_display_name", "sender"])
                    ?? ""
                let content = stringField(row, keys: ["snippet", "content", "text"])
                    ?? stringField(source, keys: ["snippet", "content", "text"])
                    ?? ""
                lines.append("[\(time)] \(sender): \(content)")
            }
        } else if let text = try? String(contentsOf: txtURL, encoding: .utf8) {
            lines.append(contentsOf: text.components(separatedBy: .newlines))
        }
        return lines
    }

    private static func fallbackASCIIPdf(lines: [String]) -> Data {
        let pageWidth = 612
        let pageHeight = 792
        let margin = 48
        let lineHeight = 14
        let linesPerPage = max(1, (pageHeight - margin * 2) / lineHeight)
        let pages = stride(from: 0, to: max(lines.count, 1), by: linesPerPage).map {
            Array(lines[$0..<min($0 + linesPerPage, lines.count)])
        }

        var objects: [Data] = []
        func add(_ s: String) { objects.append(Data(s.utf8)) }
        func add(_ d: Data) { objects.append(d) }

        add("<< /Type /Catalog /Pages 2 0 R >>\n")
        add("PLACEHOLDER")
        add("<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\n")

        var contentIds: [Int] = []
        for pageLines in pages {
            var stream = "BT /F1 10 Tf \(margin) \(pageHeight - margin) Td \(lineHeight) TL\n"
            for (i, line) in pageLines.enumerated() {
                let t = escapePDF(asciiOnly(line))
                stream += i == 0 ? "(\(t)) Tj\n" : "T* (\(t)) Tj\n"
            }
            stream += "ET"
            let body = Data(stream.utf8)
            var obj = Data("<< /Length \(body.count) >>\nstream\n".utf8)
            obj.append(body)
            obj.append(Data("\nendstream\n".utf8))
            add(obj)
            contentIds.append(objects.count)
        }

        var pageIds: [Int] = []
        for cid in contentIds {
            add("<< /Type /Page /Parent 2 0 R /MediaBox [0 0 \(pageWidth) \(pageHeight)] /Contents \(cid) 0 R /Resources << /Font << /F1 3 0 R >> >> >>\n")
            pageIds.append(objects.count)
        }
        let kids = pageIds.map { "\($0) 0 R" }.joined(separator: " ")
        objects[1] = Data("<< /Type /Pages /Kids [\(kids)] /Count \(pageIds.count) >>\n".utf8)

        var pdf = Data("%PDF-1.4\n".utf8)
        var offsets = [0]
        for (i, obj) in objects.enumerated() {
            offsets.append(pdf.count)
            pdf.append(Data("\(i + 1) 0 obj\n".utf8))
            pdf.append(obj)
            if obj.last != 0x0A { pdf.append(Data("\n".utf8)) }
            pdf.append(Data("endobj\n".utf8))
        }
        let xref = pdf.count
        pdf.append(Data("xref\n0 \(objects.count + 1)\n0000000000 65535 f \n".utf8))
        for off in offsets.dropFirst() {
            pdf.append(Data(String(format: "%010d 00000 n \n", off).utf8))
        }
        pdf.append(Data("trailer\n<< /Size \(objects.count + 1) /Root 1 0 R >>\nstartxref\n\(xref)\n%%EOF\n".utf8))
        return pdf
    }

    private static func asciiOnly(_ text: String) -> String {
        String(text.prefix(110).map { ch -> Character in
            if ch == "\\" || ch == "(" || ch == ")" { return " " }
            return ch.isASCII ? ch : "?"
        })
    }

    private static func escapePDF(_ text: String) -> String {
        text.replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "(", with: "\\(")
            .replacingOccurrences(of: ")", with: "\\)")
    }

    private static func messageRows(from root: Any) -> [[String: Any]] {
        if let array = root as? [[String: Any]] { return array }
        if let dict = root as? [String: Any] {
            for key in ["items", "messages", "results"] {
                if let items = dict[key] as? [[String: Any]] { return items }
            }
        }
        return []
    }

    private static func stringField(_ row: [String: Any], keys: [String]) -> String? {
        for key in keys {
            if let value = row[key] as? String, !value.isEmpty { return value }
        }
        return nil
    }

    private static func sanitize(_ name: String) -> String {
        let invalid = CharacterSet(charactersIn: "/\\:?%*|\"<>")
        let cleaned = name.components(separatedBy: invalid).joined(separator: "_")
        return cleaned.isEmpty ? "chat" : cleaned
    }

    private static func timestamp() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd_HHmmss"
        return f.string(from: Date())
    }
}
